include/zephyr/storage/flash_map.h

```diff
@@ -65,6 +65,19 @@ struct flash_area {
 	const struct device *fa_dev;
 };
 
+/**
+ * @brief Flash partition label
+ *
+ * This structure represents the separate flash area label
+ * seaprated from flash_area to benefit size optimization if its unused.
+ */
+struct flash_area_lbl {
+	/** ID number */
+	uint8_t fa_id;
+	/** Partition label if defined in DTS. Otherwise nullptr; */
+	const char *fa_label;
+};
```

Comment:
There is not need to define another structure for that, just add `fa_label` pointer to flash_area structure and guard it around with Kconfig that will enable its existence.
We already have a map of flash areas and there is no need to create parallel one, you can use the id in that map to pick `fa_label` from it. Both maps have to have the same number of entries basically, unless you want to skip areas without labels, and you will not save much space here. Each flash_area_lbl will need 8 bytes, while aligned to 32bit addressing, while addition of `fa_label` to `flash_area` will just take 4.

Reply:
Ok, but I was a bit concerned about adding another kConfig option and making the web of config options even more complex. Or should I just use `if CONFIG_FLASH_MAP_SHELL` for now and remove the more generic `flash_area_label` function again?

Reply:
@MauritsFassaert You can add, reasonably, as many Kconfigs as there are needed to complete your feature. Just add something like FLASH_MAP_LABELS that will be around the `fa_label` and change FLASH_MAP_SHELL to depend on it.
Leave the function as we would like to avoid non-tree code to access `flash_area` object directly, so the function should the only way to get the label out of `flash_area`, for such code.
I do not remember your function now, but I suspect it has been returning `const char *`, so you need to have it within `CONFIG_FLASH_MAP_LABELS` too, as you are not able to return `-ENOTSUP`.

Reply:
sorry I got held up by another issue in main for a few days. 

I added the FLASH_MAP_LABELS config, but a simple dependency seemed unintuitive to me. Why have two separate options if both have to be enabled manually for this shell command to work? So I decoupled them and kept the old implementation without labels column if the labels option is not used. This has the added advantage that by default the behavior of the shell is unchanged.

Reply:
You can modify `FLASH_MAP_SHELL` to depend on `FLASH_MAP_LABELS ` or `imply FLASH_MAP_LABELS `.
And you are missing copyright updates of files you have touched.

Reply:
Or if you want it to print nothing I think it is fine, you can also add #warning in shell source in case when `FLASH_MAP_LABELS` is not selected, that map will be printed without labels.

Reply:
Ok, if you prefer. If FLASH_MAP_LABELS is `n` I will print `-` as if the property is not defined, but by default it is `imply`-ed by the `FLASH_MAP_SHELL`. I reran the `rom_report` to confirm that the option still reduces the footprint with these changes.